### PR TITLE
Allow synchronous replication if specified in config.

### PR DESCRIPTION
**Pull request by poornas**
- The replication configuration is extended to add a
boolean flag for user to configure synchronous replication.The
default mode is asynchronous.

If the target node is down, falls back to healing to catch up.
## Motivation and Context
Gives user options on replication mode

## How to test this PR?
Needs https://github.com/minio/mc/pull/3473

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#d9ee9e8b-0c1b-423a-89f9-1d5ba85c7bca) |
<a name="d9ee9e8b-0c1b-423a-89f9-1d5ba85c7bca"></a>
### 10780-b82ba56/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30426
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8cea66133445:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 56 seconds
(4/15) Running aws-sdk-ruby tests ... done in 21 seconds
(5/15) Running awscli tests ... done in 2 minutes and 50 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 29 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 37 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 13 seconds
{
  "args": {
    "bucketName": "minio-go-test-jplf09zirnxocewx",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1081,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 48 seconds
(10/15) Running minio-js tests ... FAILED in 47 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 79,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 19 minutes and 14 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 21 seconds
(12/15) Running s3select tests ... done in 41 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-replication.go [hunk 1]
```diff
@@ -79,35 +79,37 @@ func validateReplicationDestination(ctx context.Context, bucket string, rCfg *re
 	return false, BucketRemoteTargetNotFound{Bucket: bucket}
 }
 
-func mustReplicateWeb(ctx context.Context, r *http.Request, bucket, object string, meta map[string]string, replStatus string, permErr APIErrorCode) bool {
+func mustReplicateWeb(ctx context.Context, r *http.Request, bucket, object string, meta map[string]string, replStatus string, permErr APIErrorCode) (replicate bool, sync bool) {
 	if permErr != ErrNone {
-		return false
+		return replicate, sync
```
Comment (kerneltime) [inline cmd/bucket-replication.go]:
```suggestion
		return
```

File: cmd/bucket-replication.go [hunk 2]
```diff
@@ -79,35 +79,37 @@ func validateReplicationDestination(ctx context.Context, bucket string, rCfg *re
...
 }
 
-// mustReplicate returns true if object meets replication criteria.
-func mustReplicate(ctx context.Context, r *http.Request, bucket, object string, meta map[string]string, replStatus string) bool {
+// mustReplicate returns 2 booleans - true if object meets replication criteria and true if replication is to be done in
+// a synchronous manner.
+func mustReplicate(ctx context.Context, r *http.Request, bucket, object string, meta map[string]string, replStatus string) (replicate bool, sync bool) {
 	if s3Err := isPutActionAllowed(getRequestAuthType(r), bucket, "", r, iampolicy.GetReplicationConfigurationAction); s3Err != ErrNone {
-		return false
+		return replicate, sync
```
Comment (kerneltime) [inline cmd/bucket-replication.go]:
```suggestion
		return
```

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10780-0bf8fd2/gopath/src/github.com/minio
+ git -C 10780-0bf8fd2/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10780-0bf8fd2/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10780-0bf8fd2/gopath/src/github.com/minio/minio fetch --quiet minio pull/10780/head:pr10780
+ git -C 10780-0bf8fd2/gopath/src/github.com/minio/minio checkout --quiet pr10780
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10780-0bf8fd2/gopath
+ make -C 10780-0bf8fd2/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/minio-go/v7@v7.0.6-0.20200929220449-755b5633803a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-0bf8fd2/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
go: github.com/minio/minio-go/v7@v7.0.6-0.20200929220449-755b5633803a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-0bf8fd2/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
make: *** [build] Error 1

```

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10780-af5726f/gopath/src/github.com/minio
+ git -C 10780-af5726f/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10780-af5726f/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10780-af5726f/gopath/src/github.com/minio/minio fetch --quiet minio pull/10780/head:pr10780
+ git -C 10780-af5726f/gopath/src/github.com/minio/minio checkout --quiet pr10780
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10780-af5726f/gopath
+ make -C 10780-af5726f/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/minio-go/v7@v7.0.6-0.20201123105958-26addf203d4a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-af5726f/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
go: github.com/minio/minio-go/v7@v7.0.6-0.20201123105958-26addf203d4a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-af5726f/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
make: *** [build] Error 1

```

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10780-7d337f6/gopath/src/github.com/minio
+ git -C 10780-7d337f6/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10780-7d337f6/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10780-7d337f6/gopath/src/github.com/minio/minio fetch --quiet minio pull/10780/head:pr10780
+ git -C 10780-7d337f6/gopath/src/github.com/minio/minio checkout --quiet pr10780
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10780-7d337f6/gopath
+ make -C 10780-7d337f6/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/minio-go/v7@v7.0.6-0.20201123105958-26addf203d4a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-7d337f6/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
go: github.com/minio/minio-go/v7@v7.0.6-0.20201123105958-26addf203d4a: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-7d337f6/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
make: *** [build] Error 1

```

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10780-d61c779/gopath/src/github.com/minio
+ git -C 10780-d61c779/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10780-d61c779/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10780-d61c779/gopath/src/github.com/minio/minio fetch --quiet minio pull/10780/head:pr10780
+ git -C 10780-d61c779/gopath/src/github.com/minio/minio checkout --quiet pr10780
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10780-d61c779/gopath
+ make -C 10780-d61c779/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/minio-go/v7@v7.0.6: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-d61c779/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
go: github.com/minio/minio-go/v7@v7.0.6: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10780-d61c779/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
make: *** [build] Error 1

```

**Comment by harshavardhana** [comment]
Other than rebase and review - what is this PR pending on @poornas ?

**Comment by harshavardhana** [comment]
> Other than rebase and review - what is this PR pending on @poornas ?

Also which user/customer requirement is this based on? - can we resubmit when there is a real requirement?

**Pull request closed without merging**